File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int tar) {
        int lo=0, hi=(int)nums.size()-1, z=-1;
        while(lo<hi){
            int mid=(lo+hi)>>1;
            if(nums[lo]<nums[hi]){
                if(nums[lo]>tar || nums[hi]<tar) return -1;
                else if(nums[mid]<tar){
                    lo=mid+1;
                }
                else hi=mid;
            }
            else if(nums[mid]>=nums[lo]){
                if(nums[mid]<tar || nums[lo]>tar){
                    lo=mid+1;
                }
                else{
                    hi=mid;
                }
            }
            else{
                if(nums[mid]<tar && tar<=nums[hi]){
                    lo=mid+1;
                }
                else hi=mid;
            }
        }
        return nums[lo]==tar?lo:-1;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    private int findMinIndex(int[] nums){
        int low= 0, high= nums.length-1;
        while(low<high){
            int mid= low+(high-low)/2;
            if(nums[mid]<nums[high]){
                high=mid;
            }
            else{
                low=mid+1;
            }
        }
        return low;
    }
    private int binarySearch(int[] nums, int lo, int hi, int target){
        int mid;
        while(lo<=hi){
            mid= lo+(hi-lo)/2;
            if(nums[mid]==target) return mid;
            else if(nums[mid]>target) hi= mid-1;
            else lo= mid+1;
        }
        return -1;
    }
    public int search(int[] nums, int target) {
        int n= nums.length;
        int ind= findMinIndex(nums);
        System.out.println(ind);
        if(nums[n-1]<target){
            return binarySearch(nums, 0, ind-1, target);
        }
        else{
            return binarySearch(nums, ind, n-1, target);
        }
    }
}